### PR TITLE
Service for getaddrinfo

### PR DESCRIPTION
**Pull request by ki11roy**
This branch is a development of #232. 
Some major changes are support of search domains and combining of ares_parse_a/aaaa_reply into ares__parse_into_addrinfo, though I haven't touched the original functions yet. 
I had to introduce additional flag ARES_AI_NOSORT to disable connecting to mock addresses while sorting. My first idea was to use ares_socket_functions, but there is no "getsockname" mock, so flag seems a better variant. 
Struct ares_addrinfo extended to keep ai_ttl and ai_cname_ttl.
The latest ares_getaddrinfo fixes are not actual anymore, because the code is removed.

### CODE REVIEW

**Comment by bradh352** [comment]
Before I do a full review on this, can you go ahead and clean up the legacy code?  Think you referred to ares_parse_a/aaaa_reply, probably those just need to be made into wrappers right?

**Comment by ki11roy** [comment]
I have tried and there is a problem. I need to add aliases to ares_addrinfo to completely replace the functions, but it seems too much, because there are already two extra fields ai_ttl and ai_cname_ttl (and ai_cname itself). So there are two separate parts which badly fit into the same struct.
I thought about moving from linked list as in classic addrinfo to more like something like good old hostent with cname, aliases, cname_ttl and a linked list of addrinfo with ttl’s. How about this?

**Comment by bradh352** [comment]
Again, I wouldn't worry about standards conformance since we're already making modifications to the struct *and* it is namespaced.  And memory these days is cheap, especially what we're talking about here.

I also think its probably silly for getaddrinfo to return technically less information than gethostbyname in any area at all.

My vote would be to make the ai_canonname an array structure holding the name and ttl

`
struct ares_addrinfo_cname {
  char *name;
  int ttl;
};

struct ares_addrinfo {
...
  struct ares_addrinfo_cname *ai_canonname;
  size_t                                       ai_canonname_len;
...
};
`

**Comment by ki11roy** [comment]
I need to fix the tests and wrap the functions, finally.

**Comment by bradh352** [comment]
Hmm, I'd think the cname + ttl struct would be an array since the recursive DNS server being queried could return multiple CNAMEs which are the aliases.  I'm not sure why you have a char **aliases at all.

We might as well save the ttl for each if we're already saving the cname for each.

File: ares.h [hunk 1]
```diff
@@ -570,15 +579,36 @@ struct ares_soa_reply {
...
 struct ares_addrinfo {
-  int                  ai_flags;
-  int                  ai_family;
-  int                  ai_socktype;
-  int                  ai_protocol;
-  ares_socklen_t       ai_addrlen;
-  char                 *ai_canonname;
-  struct sockaddr      *ai_addr;
-  struct ares_addrinfo *ai_next;
+  struct ares_addrinfo_cname     cname;
```
Comment (bradh352) [inline ares.h]: make this an array

File: ares.h [hunk 2]
```diff
@@ -570,15 +579,36 @@ struct ares_soa_reply {
...
+  char                       **aliases;
```
Comment (bradh352) [inline ares.h]: remove this

File: ares__parse_into_addrinfo.c [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+          if (status != ARES_SUCCESS)
+            break;
+
+          ares_free(hostname);
+          hostname = rr_data;
+
+          /* Take the min of the TTLs we see in the CNAME chain. */
+          if (ai->cname.ttl > rr_ttl)
+            ai->cname.ttl = rr_ttl;
+
```
Comment (bradh352) [inline ares__parse_into_addrinfo.c]: no reason to record a minimum, instead record the ttl associated with the cname entry in the array

File: ares__parse_into_addrinfo.c [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+          ares_free(ai->cname.name);
```
Comment (bradh352) [inline ares__parse_into_addrinfo.c]: No need to free prior cname, should be a new array entry

**Comment by ki11roy** [comment]
Got your point, but for the aliases I need them to populate hostent structure in ares_parse_a_reply. I see no similar structure in regular addrinfo and it is not CNAME itself, but a label of CNAME as per RFC2181 section 10.1.1. So I wanted to have labels as aliases and one value (multiple values as you suggested) as CNAME (that part is unfinished).

**Comment by ki11roy** [comment]
Fixed the tests and replaced ares_parse_a/aaaa_reply with ares__parse_into_addrinfo.

**Comment by bradh352** [comment]
I'm inclined to merge this and deal with any tweaks or fallout later.  It looks pretty good and provides a very nice implementation with more features (mainly ttls) than a native getaddrinfo which is nice.

@bagder any problems with this?

**Comment by daviddrysdale** [comment]
Looks like there's a Windows build failure with this ([AppVeyor](https://ci.appveyor.com/project/daviddrysdale/c-ares/builds/25382437)):
```
	cl.exe /nologo /MDd /D_DEBUG /Od /Zi /D_CRT_NONSTDC_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /RTCsu /DWIN32 /I. /I.. /I gmock-1.8.0 /W3 /EHsc /FD /DCARES_STATICLIB /Fo.\msvc\arestest\lib-debug\obj\ares-test-internal.obj /Fd.\msvc\arestest\lib-debug\obj\ /c .\ares-test-internal.cc
ares-test-internal.cc
C:\Program Files (x86)\Windows Kits\8.1\include\shared\winerror.h(26345) : warning C4005: 'NOERROR' : macro redefinition
        C:\projects\c-ares\nameser.h(144) : see previous definition of 'NOERROR'
.\ares-test-internal.cc(366) : error C2065: 'EnvValue' : undeclared identifier
.\ares-test-internal.cc(366) : error C2146: syntax error : missing ';' before identifier 'with_env'
.\ares-test-internal.cc(366) : error C3861: 'with_env': identifier not found
.\ares-test-internal.cc(386) : error C2065: 'EnvValue' : undeclared identifier
.\ares-test-internal.cc(386) : error C2146: syntax error : missing ';' before identifier 'with_env'
.\ares-test-internal.cc(386) : error C3861: 'with_env': identifier not found
.\ares-test-internal.cc(406) : error C2065: 'EnvValue' : undeclared identifier
.\ares-test-internal.cc(406) : error C2146: syntax error : missing ';' before identifier 'with_env'
.\ares-test-internal.cc(406) : error C3861: 'with_env': identifier not found
.\ares-test-internal.cc(426) : error C2065: 'EnvValue' : undeclared identifier
.\ares-test-internal.cc(426) : error C2146: syntax error : missing ';' before identifier 'with_env'
.\ares-test-internal.cc(426) : error C3861: 'with_env': identifier not found
.\ares-test-internal.cc(575) : warning C4267: 'argument' : conversion from 'size_t' to 'int', possible loss of data
NMAKE : fatal error
```

**Comment by daviddrysdale** [comment]
Also, I'm getting some OSS-Fuzz failure reports from this, in a couple of variants:
 - [ares_parse_a_reply.c:160](https://github.com/c-ares/c-ares/blob/master/ares_parse_a_reply.c#L160) has a WRITE stack buffer overflow &ndash; can we check all the count tests (e.g. l155) are correct?
 - [ares_parse_aaaa_reply.c:156](https://github.com/c-ares/c-ares/blob/master/ares_parse_aaaa_reply.c#L156) has a READ heap buffer overflow.

**Comment by ki11roy** [comment]
@daviddrysdale thanks! i'll check it.

**Comment by bradh352** [comment]
> Looks like there's a Windows build failure with this ([AppVeyor](https://ci.appveyor.com/project/daviddrysdale/c-ares/builds/25382437)):
> 
> ```
> 	cl.exe /nologo /MDd /D_DEBUG /Od /Zi /D_CRT_NONSTDC_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /RTCsu /DWIN32 /I. /I.. /I gmock-1.8.0 /W3 /EHsc /FD /DCARES_STATICLIB /Fo.\msvc\arestest\lib-debug\obj\ares-test-internal.obj /Fd.\msvc\arestest\lib-debug\obj\ /c .\ares-test-internal.cc
> ares-test-internal.cc
> C:\Program Files (x86)\Windows Kits\8.1\include\shared\winerror.h(26345) : warning C4005: 'NOERROR' : macro redefinition
>         C:\projects\c-ares\nameser.h(144) : see previous definition of 'NOERROR'
> .\ares-test-internal.cc(366) : error C2065: 'EnvValue' : undeclared identifier
> .\ares-test-internal.cc(366) : error C2146: syntax error : missing ';' before identifier 'with_env'
> .\ares-test-internal.cc(366) : error C3861: 'with_env': identifier not found
> .\ares-test-internal.cc(386) : error C2065: 'EnvValue' : undeclared identifier
> .\ares-test-internal.cc(386) : error C2146: syntax error : missing ';' before identifier 'with_env'
> .\ares-test-internal.cc(386) : error C3861: 'with_env': identifier not found
> .\ares-test-internal.cc(406) : error C2065: 'EnvValue' : undeclared identifier
> .\ares-test-internal.cc(406) : error C2146: syntax error : missing ';' before identifier 'with_env'
> .\ares-test-internal.cc(406) : error C3861: 'with_env': identifier not found
> .\ares-test-internal.cc(426) : error C2065: 'EnvValue' : undeclared identifier
> .\ares-test-internal.cc(426) : error C2146: syntax error : missing ';' before identifier 'with_env'
> .\ares-test-internal.cc(426) : error C3861: 'with_env': identifier not found
> .\ares-test-internal.cc(575) : warning C4267: 'argument' : conversion from 'size_t' to 'int', possible loss of data
> NMAKE : fatal error
> ```

Looks like EnvValue is conditionalized against Windows for some reason.  I've removed that condition and fixed the build.

**Comment by ki11roy** [comment]
> Also, I'm getting some OSS-Fuzz failure reports from this, in a couple of variants:
> 
> * [ares_parse_a_reply.c:160](https://github.com/c-ares/c-ares/blob/master/ares_parse_a_reply.c#L160) has a WRITE stack buffer overflow – can we check all the count tests (e.g. l155) are correct?
> * [ares_parse_aaaa_reply.c:156](https://github.com/c-ares/c-ares/blob/master/ares_parse_aaaa_reply.c#L156) has a READ heap buffer overflow.

#262 should fix this one

**Comment by ChristianAmmer** [comment]
Adding "Service for ares_getaddrinfo" also changed the order search domains for T_A and T_AAAA requests are tried (If I'm not wrong).

Before:
T_AAAA for host.suffix1 and parallel T_A for host.suffix1
T_AAAA for host.suffix2 and parallel T_A for host.suffix2

Now:
T_AAAA for host.suffix1
T_AAAA for host.suffix2
T_A for host.suffix1
T_A for host.suffix2

This slows down the entire lookup. In comparison to this, glibc uses parallel searches with the intention of not slowing down.

**Comment by bradh352** [comment]
@ChristianAmmer I agree that queries for ipv4 & ipv6 should be done in parallel

**Comment by ki11roy** [comment]
As far as I remember the initial purpose of that code was to fixing the incorrect behavior, which was done in mostly invasive way. So I just fixed it the other way. Everything above this is an optimization. I do agree with the statement about the parallel queries, but I my opinion the first step is fixing the code (its done) and the second is implementing the optimization. I will think on how to return it in the most painless way.

**Comment by ChristianAmmer** [comment]
Hi @bradh352 @ki11roy
I was working on a related issue today, this remembered me on this one. To not forget it, I created #285.

**Pull request merged**